### PR TITLE
Rails OAuth Sample for Veteran Health APIs

### PR DESCRIPTION
**Pull request by erluti**
this is a sample app to demonstrate OAuth flow and Veteran Health API responses
* login and callback pages show information transferred between app and authentication server
* api requests show raw json response and navigable links through the API

For this issue: https://app.zenhub.com/workspaces/vets-contrib-5ac272fff41bfa50c0e46c48/issues/department-of-veterans-affairs/vets-contrib/1830

### CODE REVIEW

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+  def login
+    # create a linke to the oauth server based on the "Authorization Code Flow" described here https://developer.va.gov/explore/verification/docs/authorization
+    nonce_base = SecureRandom.base64(20)
+    session[:id] = nil
+    session[:nonce_key] = nonce_base
+    session[:login_time] = Time.zone.now.to_i
+    session[:oauth_response] = nil
+    session[:oauth_code] = nil
+    scope = 'openid profile launch/patient patient/Patient.read patient/AllergyIntolerance.read patient/Condition.read patient/DiagnosticReport.read patient/Immunization.read patient/Medication.read patient/MedicationOrder.read patient/MedicationStatement.read patient/Observation.read patient/Procedure.read'
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: Maybe throw the scope in a method for a little better readability

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
+class SessionController < ApplicationController
+  before_action :require_auth, only: :show
+
+  def login
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: Does this method build a url and then return that url? If so, the name "login" for this method is confusing. Maybe something like "generate_oath_url" ?

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      redirect_to(login_path) and return
+    end
+    @verified_state = params[:state].to_i == session[:login_time]
+    @body = {
+      code: params[:code],
+      state: params[:state],
+      grant_type: 'authorization_code',
+      redirect_uri: 'http://localhost:3000/callback'
+    }
+    @post_url = 'https://dev-api.va.gov/oauth2/token'
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: Maybe put this in a private method. This var being set here is confusing because it isn't immediately used

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    auth = { username: ENV['va_developer_client_id'], password: ENV['va_developer_client_secret'] }
+    @auth = auth.dup
+    @auth[:password] = "<Client Secret found in application.yml>"
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: I think this block setting auth stuff can be abstracted into a method as well. And perhaps add a small comment about the password obfuscation, it took me a minute to read around and figure that out

**Comment by dhall** [comment]
The app says veteran verification in multiple spots .. guessing copy/paste issue?
Screen Shot 2019-05-03 at 8 39 36 AM

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            attributes = Session.attributes_from_oauth(oauth_post)
+            @session = Session.new(attributes)
+          end
+          session[:oauth_response]
+        else
+          # session[:oauth_response] is an array of keys, so get data from saved session
+          @session = Session.find(session[:id])
+          hide_sensitive_data(@session.attributes.reject { |k,v| %w(id created_at updated_at).include?(k) })
+        end
+      @session.validate_session(session) if @session
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: I am pretty confused about what the difference is between `@session` and `session` is, so this line hurt my head a little

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    @auth = auth.dup
+    @auth[:password] = "<Client Secret found in application.yml>"
+    @session = nil
+
+    # set oauth_response and oauth_response_code
+    if session[:oauth_code]
+      # we are reloading the page so we don't want to make a stale request
+      @oauth_response_code = session[:oauth_code]
+      @oauth_response = 
+        if session[:oauth_response].is_a? Hash
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: This section would be more readable if you put the `oauth_response` logic in a separate method

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    @session =
+      if params[:id].blank?
+        Session.find(session[:id])
+      else
+        Session.find(params[:id])
+      end
+  end
+
+private
+  def hide_sensitive_data(response_hash)
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: Could you explain why we are hiding this data in the response?

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        sesh.expires_at = Time.zone.now
+        sesh.save!
+      end
+    end
+    redirect_to login_path
+  end
+
+  def show
+    @session =
+      if params[:id].blank?
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: I think this if block is a good candidate for a ternary, since you're assigning the result. `@session = params[:id].blank? ? Session.find(session[:id]) : Session.find(params[:id])`

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      @page, @count = page, count
+      search_param = (api_name == 'Patient' ? '_id' : 'patient')
+      @target = "https://dev-api.va.gov/services/argonaut/v0/#{api_name}?#{search_param}=#{id}&page=#{page}&_count=#{count}"
+    elsif action == :read
+      @page, @count = nil, nil
+      @target = "https://dev-api.va.gov/services/argonaut/v0/#{api_name}/#{id}"
+    else
+      raise "action must be one of the symbols :search or :read"
+    end
+    @api_response = HTTParty.get(@target, headers: { Authorization: "Bearer #{access_token}" })
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: Please add line breaks before and after your if blocks (makes it a little easier to read)

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  def code
+    @api_response.code
+  end
+
+  def response_string
+    return @response_string if @response_string
+    # use JSON pretty_generate to add readable whitespace and gsub api links to be links with the test app
+    @response_string = JSON.pretty_generate(@api_response.to_h).gsub(/\"(https:\/\/dev-api.va.gov\/services\/argonaut\/v0)\/(\w+)\/(.+)\"/, "<a href=\"/health_api/api_response/\\2/\\3\">\\1\/\\2\/\\3</a>")
+    @response_string
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: I think you can eliminate this line, and using line 25 as the return value should Just Work. I think.

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    end
+    @api_response = HTTParty.get(@target, headers: { Authorization: "Bearer #{access_token}" })
+  end
+
+  def code
+    @api_response.code
+  end
+
+  def response_string
+    return @response_string if @response_string
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: How is `@response_string` originally being set? I don't see it anywhere else

File: samples/oauth_rails_vethealth/app/models/session.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  def validate_session(session)
+    @session_errors = []
+    if expired?
+      @session_errors << { type: :expired, message: 'The session has expired.' }
+    end
+    unless authentic?(session)
+      @session_errors << { type: :inauthentic, message: 'The token is not authentic!' }
+    end
+    @validated = true
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/session.rb]: Put line breaks around the `if` and `unless` blocks so people don't mistake those for the same tree

File: samples/oauth_rails_vethealth/config/credentials.yml.enc [hunk 13]
```diff
@@ -0,0 +1 @@
+wlILgaX5F1IR89EnySh2OFh6J8Q/mKYnlnxBbw0tc3yCHH0B4PD4DY8PIqPsUiV22w85TVBHdFkoeE1aN60BqySJjyrMX2J3FiTAZfRbbzp72x19UmHGbzWVkZ1pEKQaxV9hzNnqdr62Pn4+G+nGNSjW+lv45sgqm/q3IAG7ocj4WzQPOCEZozcceyXuC1lsZGEbRyYwvN9wj1YEGfxRsa0XazfOB3AGbA48vLIqe29IFZXrgBpdmT6W32sWorYQP3mYCC2NMSwwgFb/Ik75PwDaOYnGIbK/VMFP1L8H5TUr5ZAGScbPaG2ZsF/KxCo2ApgzYDgBQNG8o/vGaWB8IyFcJXF5rVHke2mT8S6rUt4POLGnlQZDyySDRrUFLsWQ3YpKgtFFMx6oIt4Q7LsnDhQ40NQc5NTzpJFH--7oPcVmVOqfnLEgPK--byWMu4pml0Rm4oyau6I1tw==
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/config/credentials.yml.enc]: is that anyone's super secret token...?

File: samples/oauth_rails_vethealth/test/controllers/session_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class SessionControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/test/controllers/session_controller_test.rb]: Some tests for the controllers would be good. You have some special cases for the session info and stuff (I think?) that would nice to capture here

File: samples/oauth_rails_vethealth/config/credentials.yml.enc (hunk 13 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/config/credentials.yml.enc]: I was messing with this https://medium.com/cedarcode/rails-5-2-credentials-9b3324851336 and I think I inadvertently created this file!  I'm going to delete it.

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 11 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: I tend to memoize complex methods so the logic doesn't run every time a method is called.  This method was really complex when I added it.  It's still doing a lot of string substitution and stuff, so I figured I'd leave it in.

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 10 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: yup!  thanks!

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 11 above)
Comment (edmkitty) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: In that case, this entire method can be one line:
`@response_string ||= JSON.pretty_generate(@api_response.to_h).gsub(/\"(https:\/\/dev-api.va.gov\/services\/argonaut\/v0)\/(\w+)\/(.+)\"/, "<a href=\"/health_api/api_response/\\2/\\3\">\\1\/\\2\/\\3</a>")`

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb (hunk 8 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: I strongly dislike ternaries as a style preference.

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb (hunk 7 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: The `access_token` because if this was captured someone could impersonate requests to the APIs, so I decided to hide it as a best practice sort of thing.  I went back and forth on wether or not to show the JWT.  It's got the name in there so it would be PII (if it wasn't a test user), but I think I ended up showing it on the `session#show` page.

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb (hunk 8 above)
Comment (patrickvinograd) [reply samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: Can we set up rubocop with a reasonable set of rules (e.g. those used by vets-api) to avoid stylistic debates? (There are some that will still be up to discretion but in general, I favor letting the linter arbitrate most of these).

**Comment by erluti** [comment]
the database has been remade.  so delete `db/development.sqlite3` and run `rails db:migrate` again to use this locally if you've already pulled this repo

still some more changes incoming however

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb (hunk 5 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: renamed `Session` model to `Authentication`

**Comment by erluti** [comment]
Done making changes, ready for fresh reviews

File: samples/oauth_rails_vethealth/app/mailers/application_mailer.rb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: 'from@example.com'
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/mailers/application_mailer.rb]: Is this used anywhere, and if so should this be changed? Or is this just default boilerplate stuff?

File: samples/oauth_rails_vethealth/app/models/authentication.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  belongs_to :oauth_callback, optional: true
+
+  attr_reader :session_errors
+
+  def self.generate_nonce(base)
+    Digest::SHA256.hexdigest(base + ENV['va_developer_client_secret'])
+  end
+
+  def self.attributes_from_oauth(response)
+    attributes_array = response.map do |key,value|
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/authentication.rb]: If you are only replacing the 'expires_at' value, then I think mapping all of the key value pairs is excessive. Perhaps do this instead:
`attributes_array = response` (I suppose you could dup here if that's an issue)
`attributes_array[:expires_at] = Time.zone.at(response[:expires_at])`

File: samples/oauth_rails_vethealth/app/models/authentication.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+  def validate_session(session)
+    @authentication_errors = []
+    if expired?
+      @authentication_errors << { type: :expired, message: 'The session has expired.' }
+    end
+    unless authentic?(session)
+      @authentication_errors << { type: :inauthentic, message: 'The token is not authentic!' }
+    end
+    @validated = true
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/authentication.rb]: If errors are added to the `authenticated_errors` array, is this supposed to be `false` ?

File: samples/oauth_rails_vethealth/app/models/authentication.rb [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    @authentication_errors.empty?
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/authentication.rb]: I'm not sure what you mean by this code. Does it mean that the validity of the session depends on whether or not there are errors? If so, what is the purpose of the `@validated` variable? Couldn't you just call `validate_session` from within `valid_session?` ?

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 19]
```diff
@@ -0,0 +1,55 @@
+class HealthApiResponse
+  attr_reader :api_name, :target, :page, :count
+  def initialize(api_name:, id:, access_token:, action: :read, page: 1, count: 10)
+    @api_name, @id, @action = api_name, id, action
+
+    case action
+    when :search
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: Why are the actions being managed in the model and not in the controller? You could have the controller handle this logic and then call the appropriate queries on the model

File: samples/oauth_rails_vethealth/app/mailers/application_mailer.rb (hunk 15 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/mailers/application_mailer.rb]: this is boilerplate stuff.  I'm not using mailers anywhere

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    @api_response = HTTParty.get(@target, headers: { Authorization: "Bearer #{access_token}" })
+  end
+
+  def code
+    @api_response.code
+  end
+
+  def response_string
+    # use JSON pretty_generate to add readable whitespace and gsub api links to be links with the test app
+    @response_string ||= JSON.pretty_generate(@api_response.to_h).gsub(/\"(https:\/\/dev-api.va.gov\/services\/argonaut\/v0)\/(\w+)\/(.+)\"/, "<a href=\"/health_api/api_response/\\2/\\3\">\\1\/\\2\/\\3</a>")
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: For a little more readability, you could put the regex in a method too. ALL THE METHODS

File: samples/oauth_rails_vethealth/app/models/authentication.rb (hunk 17 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/authentication.rb]: `@validated` is if validation has been run at all

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+  end
+
+  def api_navs
+    return @api_navs if @api_navs
+    return nil unless (@action == :search && code == 200)
+    @api_navs = []
+    self_nav = nil
+    @api_response['link'].each do |link|
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/health_api_response.rb]: I think you could probably memoize `@api_navs` if you moved this logic into Yet Another Method, but I don't have strong feelings about it

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb [hunk 22]
```diff
@@ -0,0 +1,47 @@
+class OauthCallback < ApplicationRecord
+  serialize :response_body_raw
+
+  has_one :authentication
+
+  def auth(protect_secrets=true)
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: Could you explain when/how this would be false?

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+    pass =
+      if protect_secrets
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: YOU COULD USE A TERNARY HERE TROLLOLOLOL.

File: samples/oauth_rails_vethealth/app/assets/javascripts/health_api.coffee [hunk 24]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (dvogel) [inline samples/oauth_rails_vethealth/app/assets/javascripts/health_api.coffee]: Dropping the coffeescript dependency would be nice. As a sample app, there's a lot of value in this being as easy to approach as possible.

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+  def fetch_access_token!(session)
+    oauth_post = HTTParty.post(self.oauth_url, { basic_auth: auth(false), body: post_body })
+    self.response_body_raw = oauth_post.to_h
+    self.response_code = oauth_post.code
+    save!
+
+    if oauth_post.ok?
+      attributes = Authentication.attributes_from_oauth(oauth_post)
+      self.build_authentication(attributes)
+      if self.authentication.validate_session(session)
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: Does this code not run when you call `.save!` on the Authentication instance? Shouldn't it?

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 21 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: i believe it is memoized by returning @api_navs a couple lines up

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 19 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: I'm not sure what you mean here?  The responses are all used the same, but the target is generated differently based on whether you want to make a "read" request or a "search" request, so I thought that was handled neatly in the initializer.

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+        self.authentication.save!
+      end
+    end
+  end
+
+  def response_body
+    return @response_body if @response_body
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: You can simplify this method a bit by doing this:
```  def response_body
    @response_body ||= response_body_raw.dup.tap do |body|
      # the access_token allows requests to be made in the logged-in users name, so don't expose it to the web
      body['access_token'] = "<secret token for accessing API>"
    end
  end```

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb (hunk 25 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: I went back and forth mentally on this being a proper validation.  But since it's validating if the cookie is valid for the saved authentication instead of the session being a property of the Authentication, I decided it wasn't a validation in the rails-sense.  But I don't know a better word than "validate" to use here?

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb (hunk 22 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: I set it to false when I want to use the real values in the authentication call.
https://github.com/department-of-veterans-affairs/vets-api-clients/blob/6328c494404351c2d33e4c3afba2211947506f95/samples/oauth_rails_vethealth/app/models/oauth_callback.rb#L26
But when displaying `auth` in a view it will default to hiding the secrets

File: samples/oauth_rails_vethealth/app/controllers/session_controller.rb [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+    oauth = OauthCallback.create!(
+      verified_state: params[:state].to_i == session[:login_time],
+      code: params[:code],
+      state: params[:state],
+      oauth_url: 'https://dev-api.va.gov/oauth2/token'
+    )
+    oauth.fetch_access_token!(session)
+    session[:id] = oauth.authentication.id if oauth.authentication
+    redirect_to oauth_callback_path(oauth)
+  end
```
Comment (dvogel) [inline samples/oauth_rails_vethealth/app/controllers/session_controller.rb]: What is our rationale for handling the oauth callbacks directly rather than using a common library such as OmniAuth? Have we discussed this?

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb (hunk 26 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: I generally don't like `tap` but I think it's pretty clear here, thanks

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+
+  def auth(protect_secrets=true)
+    pass =
+      if protect_secrets
+        '<shhhhhhh>'
+      else
+        ENV['va_developer_client_secret']
+      end
+    { username: ENV['va_developer_client_id'], password: pass }
+  end
```
Comment (dvogel) [inline samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: How will the `protect_secrets == true` path be useful? Won't that always result in a failed request for the access token?

Comment (erluti) [reply samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: This app took an approach to be "educational" in that using the app will walk you through what's going on behind the scenes.  So in the views it shows the `auth` information and I didn't want to expose the client secret.  But in the actual request it's set to false because we want to use the client secret.

File: samples/oauth_rails_vethealth/app/assets/javascripts/health_api.coffee (hunk 24 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/assets/javascripts/health_api.coffee]: Those were generated by rails, but since I'm not using them I'll delete them

**Comment by edmkitty** [comment]
Great job on tests, very detailed!

File: samples/oauth_rails_vethealth/app/models/oauth_callback.rb (hunk 25 above)
Comment (edmkitty) [reply samples/oauth_rails_vethealth/app/models/oauth_callback.rb]: Ahh, I see. Maybe just add some comments about it.

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 21 above)
Comment (edmkitty) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: Oh yeah, I just meant in the sense of using the handy ruby `||=` operator, but it doesn't make a difference really

File: samples/oauth_rails_vethealth/app/models/health_api_response.rb (hunk 19 above)
Comment (edmkitty) [reply samples/oauth_rails_vethealth/app/models/health_api_response.rb]: Discussed, concluded that refactoring is too painful here

File: samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb [hunk 29]
```diff
@@ -0,0 +1,7 @@
+<h1 class="display-2">Callback <em>Failed</em></h1>
+<p class="lead">
+    The time of login was sent in the "state" parameter, but the returned state,
```
Comment (rtravitz) [inline samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb]: Found an interesting edge case that gets this to show. When I start the flow by clicking "Oauth Login", I go to ID.me. Invariably, I have forgotten to copy the test user's email, so I open the sample app back up in another tab. This seems to reset the state to a new time. When I then go back to the first tab and continue with the flow, I eventually get to the attached view.

Two questions:
1. Is there a way to separate the state that gets sent to account for another tab being opened?
2. Should we change the text that says "Now with the user identify verified... etc." when there is an error case? It is a little confusing because the error message underneath implies the test user was in fact *not* verified.
Screen Shot 2019-05-14 at 11 38 09 AM

Comment (erluti) [reply samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb]: 1 - the state i'm trying to match is embedded into the "login with oauth" link.  I reset that state stored in the session on the login page.  it's a cookie so i'm not sure there's a simple way to split that between tabs. i suppose i could not reset the state it if it was generated within the last couple minutes?

2 - Yeah, I can make that page clearer.  I didn't expect you to find it, honestly!

Comment (erluti) [reply samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb]: So this screenshot is from an old version of the code.  Pull the latest and the error page here is clearer.  I'm still thinking through keeping the state the same for a few minutes before creatinga . new one.

File: samples/oauth_rails_vethealth/app/views/oauth_callbacks/show.html.erb [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+    </tbody>
+</table>
+<% if @oauth_callback.authentication %>
+    <% if @oauth_callback.authentication.authentication_errors.empty? %>
+        <p>Authentication was saved with ID <code><%= @oauth_callback.authentication.id %></code>.
+        Use this <%= link_to 'authentication record', authentication_path(@oauth_callback.authentication.id) %> to set headers for requests to the API.</p>
+    <% else %>
+        <p>The session does not match the oauth response.  These are the issues:</p>
+        <ul>
+        <% @oauth_callback.authentication.session_errors.each do |error| %>
```
Comment (edmkitty) [inline samples/oauth_rails_vethealth/app/views/oauth_callbacks/show.html.erb]: This should be `authentication_errors` now, right?

Comment (erluti) [reply samples/oauth_rails_vethealth/app/views/oauth_callbacks/show.html.erb]: thanks!  good catch!

File: samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb (hunk 29 above)
Comment (erluti) [reply samples/oauth_rails_vethealth/app/views/oauth_callbacks/show_unverified_state.html.erb]: So I fixed 1 by using logout to cleanup session information instead of just resetting it on the login page

**Comment by edmkitty** [comment]
LGTM! :shipit: 🚀
